content/posts/introducing-delouse.md

```diff
@@ -0,0 +1,158 @@
+---
+title: 'Introducing delouse'
+excerpt: |
+    Introduction post on our internal rust tokio async debugging tool, delouse!
+coverImage: 'documentation-assets/introducing-delouse.jpg'
+date: '2024-01-08T15:00:00Z'
+author:
+    name: Paul Tagliamonte
+    picture: '/documentation-assets/paultag.jpg'
+ogImage:
+    url: 'documentation-assets/introducing-delouse.jpg'
+---
+
+At Zoo, most of our user-facing stack is written in Rust, where we’ve
+wholeheartedly adopted the [tokio](https://tokio.rs/) async ecosystem of
+libraries - a huge win in terms of library ecosystem and server performance.
+With that comes a bit of overhead, especially around the still nascent async
+Rust tooling and developer experience. We’ve found debugging async Rust
+services can be tricky, since a lot of the Rust ecosystem is split along the
+async/sync boundary, and loads of async specific tools are still experimental.
+
+A few months ago, we fixed an issue with the
+[Design API](https://zoo.dev/design-api)
+where we would notice an intermittent crash due to file descriptor exhaustion
+after enough load. On a hunch, I exported `tokio` task counts via our metrics
+stack using part of the
+[unstable API surface](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.metrics),
+and determined there was an async task leak. Tasks would spawn, but never shut
+down on their own after the WebSocket closed because they collectively
+maintained ownership of enough objects to not drop their peer’s channel
+handles. This issue went unnoticed for months because async tasks (like many
+other green threads) are inexpensive in terms of CPU – they don’t wake up all
+that often and in our case, we didn't see OOMs or excessive RAM usage due to
```

Comment:
I've never heard of OOMs. Would it be worth adding a Glossary entry for it, and adding a link here? Although I know that this post's audience is more technical.

Reply:
Ah! They're a term of art for the "out of memory" handler -- the oomkiller. Saying OOM is totally redundant here so I removed it, since it's just more words between the text and the point